pkg/grpc/ratelimit/ratelimit_impl.go

```diff
@@ -0,0 +1,104 @@
+package ratelimit
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/ratelimit"
+	"github.com/stackrox/rox/pkg/httputil"
+	"github.com/stackrox/rox/pkg/sync"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+)
+
+type rateLimiter struct {
+	mutex              sync.Mutex
+	tokenBucketLimiter *rate.Limiter
+}
+
+// Limit implements "ratelimit.Limiter" interface.
+func (limiter *rateLimiter) Limit() bool {
+	return !limiter.tokenBucketLimiter.Allow()
+}
+
+func (limiter *rateLimiter) modifyRateLimit(limitDelta int) {
+	if limiter.tokenBucketLimiter.Limit() == rate.Inf {
+		return
+	}
+
+	limiter.mutex.Lock()
+	defer limiter.mutex.Unlock()
```

Comment:
Do we need this mutex? I understand that it's required to prevent concurrent burst limit modifications but maybe it could be move to `if` body?

Reply:
We have to lock the calculated value for `newBurst`. If we would move it into `if` it could lead to wrong values. 

Imagine two routines g1 and g2. And this execution flow:
```
g1 -> newBurst = 20 + 10 (no lock yet)
-> context switch g1 -> sleep, g2 -> continue
g2 -> newBurst = 20 +10 (no lock yet)
g2 -> if { mutex + change burst to 30 }
-> context switch g2 -> sleep, g1 -> continue
g1 -> if { mutex + change burst to 30 }
```
The outcome would be 30 instead of 40 which would be expected from two increases.